### DISCUSSION TITLE
Best practices when sharing BE and FE code to avoid un tree shakable imports

### DISCUSSION THREAD

**Discussion by athammer** [webpack/webpack#16412] [Q&A]
Hello!

I have a web app that has a util folder full of code that is shared between the server and the front-end. The problem starts when a FE file imports a file from the that util folder. The FE JS bundle is now 2MBs+ because all of the imports/packages in that util file are now packaged into the FE JS bundle even if I'm just importing a single variable on the FE.  I'm very careful not to bring in any packages on the FE that will increase the bundle but obviously I don't care as much on the BE. 

They won't automatically tree shake as webpack doesn't know if they only contain pure functions (to my knowledge).

For visual people

`utils/math.js`
```
import moment from "moment";

export function square(x) {
    return x * x;
}

export function cube(x) {
    return moment.format(x * x * x);
}
```

`Welcome.tsx` (React file)
```
import { square } from "/utils/math.js";

const Welcome = (props) => {
...
```

FE JS bundle now bundles moment.js even though we're not using it.

It's also been described [here](https://stackoverflow.com/questions/67503926/webpack-tree-shaking-not-working-when-importing-from-a-file-that-has-other-impor) and the solution seems to be marking every package as side effect free which you can imagine is a nightmare and would lead to some nasty bugs. I've also read the [tree-shaking docs ](https://webpack.js.org/guides/tree-shaking/)and had no luck. I was running in production mode as well.

The only solution I've come up with is to make FE only files that do not have any BE packages that could blow up the bundle size which is almost as bad. How do others do this?

Thank you!

**Comment by athammer** [comment]
tsconfig is
```
{
	"$schema": "https://json.schemastore.org/tsconfig",
	"display": "Next.js",

	"compilerOptions": {
		"allowJs": true,
		"composite": false,
		"declaration": true,
		"esModuleInterop": true,
		"forceConsistentCasingInFileNames": true,
		"incremental": true,
		"inlineSources": false,
		"isolatedModules": true,
		"jsx": "preserve",
		"lib": ["dom", "dom.iterable", "esnext"],
		"moduleResolution": "node",
		"module": "esnext",
		"noEmit": true,
		"noUnusedLocals": false,
		"noUnusedParameters": false,
		"preserveWatchOutput": true,
		"resolveJsonModule": true,
		"skipLibCheck": true,
		"strict": true,
		"target": "es6",
		"baseUrl": "./src",
		"paths": {
			"~components/*": ["components/*"],
			"~features/*": ["features/*"],
			"~pages/*": ["pages/*"],
			"~server/*": ["server/*"],
			"~styles/*": ["styles/*"],
			"~types/*": ["types/*"],
			"~locales/*": ["locales/*"],
			"~utils/*": ["utils/*"],
			"~layouts/*": ["layouts/*"],
			"~public/*": ["./../public/*"]
		}
	},
	"include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
	"exclude": ["node_modules"]
}

```

in case anyone wants it

**Reply by alexander-akait** [reply]
For moment you can try https://webpack.js.org/plugins/context-replacement-plugin/#content-callback, anyway you need to mark side effect free where you don't have it, otherwise we can't undestand, also you can try to use dynamic import, so you will not load the bug code until you don't need it

**Reply by athammer** [reply]
So if I have moment at the top of the file as an import we'd mark it as sideEffect free by doing something like `import moment /*#__PURE__*/ from 'moment'`? I tried putting /*#__PURE__*/ in different places and also some things the docs suggested but couldn't get it to work sadly.

Thank you!

**Reply by alexander-akait** [reply]
You can find more annotation examples here https://github.com/terser/terser#annotations, also you can override `sifeEffects` on the loader level https://webpack.js.org/configuration/module/#rulesideeffects using regexp
